### PR TITLE
Multi-cursor: slice predicate action

### PR DESCRIPTION
**Pull request by yycptt**
**What changed?**
- Add slice predicate action which will be run upon checkpointing queue state

**Why?**
- slicePredicateAction will move all slices in default reader, with non-universal predicate to the next reader so that upon restart task loading for those slices won't blocking loading for other slices in the default reader.
- Slice.ShrinkScope() will shrink its predicate when there's few namespaces left. But those slices may still in the default reader. If there are many such slices in the default reader, then upon restart task loading for other namespace will be blocked. So we need to move those slices to a different reader.

**How did you test it?**
- Unit test

**Potential risks**

**Is hotfix candidate?**

### CODE REVIEW

File: service/history/tasks/predicates.go [hunk 1]
```diff
@@ -150,6 +150,21 @@ func OrPredicates(a Predicate, b Predicate) Predicate {
 	return predicates.Or(a, b)
 }
 
+func IsUniverisalPredicate(p Predicate) bool {
+	_, ok := p.(*predicates.UniversalImpl[Task])
```
Comment (yiminc) [inline service/history/tasks/predicates.go:154]: UniversalImpl does not really need type parameter.

File: service/history/queues/action_slice_predicate.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+		return
+	}
+
+	if a.maxReaderCount <= DefaultReaderId+1 {
+		return
+	}
+
+	sliceCount := a.monitor.GetSliceCount(DefaultReaderId)
+	pendingTasks := 0
+	hasNonUniversalPredicate := false
```
Comment (yiminc) [inline service/history/queues/action_slice_predicate.go:77]: would it be easer to read if you revert the logic, e.g. define hasUniversalPredicate here.

**Review by yiminc — APPROVED** [review]

File: service/history/tasks/predicates.go (hunk 1 above)
Comment (yycptt) [reply service/history/tasks/predicates.go:154]: Logically yes. But `UniversalImpl` needs to have the `Test(T) bool` method and golang currently doesn't allow parameterized method.

https://go.googlesource.com/proposal/+/refs/heads/master/design/43651-type-parameters.md#no-parameterized-methods

File: service/history/queues/action_slice_predicate.go (hunk 2 above)
Comment (yycptt) [reply service/history/queues/action_slice_predicate.go:77]: I am not sure if I follow. The goal is to know if there's any slice with non-universal predicate in the reader, and proceed only when there is at least one. 

It doesn't really care about whether or not there's any slice with universal predicate. So `hasUniversalPredicate` will be useless here. Unless I count # of slice with universal predicate and compare to total slices, but that seems more complex.

**Pull request merged**